### PR TITLE
Fetch homeUrl from .well-known

### PR DESCRIPTION
**Pull request by johannes-krude**
A custom home.html can be used to provide users with important information and rules. When users install element-desktop for a more permanent setup, the custom home.html is not shown anymore.

This commit introduces fetching of a custom homeUrl from `.well-known/matrix/client["io.element.embedded_pages"]["home_url"]`. As a result the home.html is also shown to element-desktop users.

I guess, this feature should also be documented in vector-im/element-web.

Notes: Configure custom home.html via `.well-known/matrix/client["io.element.embedded_pages"]["home_url"]` for all your element-web/desktop users

Signed-off-by: Johannes Krude <johannes.krude@comsys.rwth-aachen.de>

Type: enhancement

---
Here's what your changelog entry will look like:

## ✨ Features
 * Configure custom home.html via `.well-known/matrix/client["io.element.embedded_pages"]["home_url"]` for all your element-web/desktop users ([\#7790](https://github.com/matrix-org/matrix-react-sdk/pull/7790)). Contributed by @johannes-krude.

### CODE REVIEW

**Review by t3chguy — APPROVED** [review]
Looks sane code-wise - needs documentation in element-web's repo somewhere

**Comment by johannes-krude** [comment]
The requested documentation has a separate PR https://github.com/vector-im/element-web/pull/21066.

Is there anything else missing for this PR to get a proper review and move forward?

**Review by kittykat — CHANGES_REQUESTED** [review]
Makes sense to have fallbacks from a product pov. For prioritisation, ideally it would be:
1. Use home_url from config.json if it exists
2. Check if server has a home_url
3. Use the built in default HomePage

Our use case for preferring config.json over server is in cases where we want users on a specific app to have a specific experience. For example, what we do for FOSDEM where we discourage accounts on the homeserver, but want to serve the same home page to all users to provide a high quality experience for the conference.

**Review by t3chguy — APPROVED** [review]
Code-wise LGTM

File: src/utils/pages.ts [hunk 1]
```diff
@@ -15,11 +15,16 @@ limitations under the License.
...
 
 import { ConfigOptions } from "../SdkConfig";
+import { getEmbeddedPagesWellKnown } from '../utils/WellKnownUtils';
 
 export function getHomePageUrl(appConfig: ConfigOptions): string | null {
     const pagesConfig = appConfig.embeddedPages;
     let pageUrl = pagesConfig?.homeUrl;
 
+    if (!pageUrl) {
+        pageUrl = getEmbeddedPagesWellKnown()?.home_url;
```
Comment (turt2live) [inline src/utils/pages.ts:25]: this should be after the config fallback

**Review by turt2live — CHANGES_REQUESTED** [review]

File: src/utils/pages.ts (hunk 1 above)
Comment (johannes-krude) [reply src/utils/pages.ts:25]: @kittykat explicitly requested in https://github.com/matrix-org/matrix-react-sdk/pull/7790#pullrequestreview-968105088 to prefer config.json and only to use .well-known if config.json does not have a homeUrl.

Comment (turt2live) [reply src/utils/pages.ts:25]: the statement below is more config.json though?

Comment (johannes-krude) [reply src/utils/pages.ts]: You are correct. It is now fixed.

**Comment by t3chguy** [comment]
@johannes-krude please fix the merge conflict

**Review by turt2live — APPROVED** [review]

**Comment by johannes-krude** [comment]
I merged this PR with develop and tested it.

> Makes sense to have fallbacks from a product pov. For prioritisation, ideally it would be:
> 
>     1. Use home_url from config.json if it exists
> 
>     2. Check if server has a home_url
> 
>     3. Use the built in default HomePage

`getHomePageUrl` now does the following:
1. return `config.json["embeddedPages"]["homeUrl"]` if it exists
2. return the deprecated `config.json["welcomePageUrl"]` if it exists
3. return `.well-known/matrix/client["io.element.embedded_pages"]["home_url"]` if it exists
4. return `null` which results in using the built in default HomePage

**Review by kittykat — APPROVED** [review]

File: src/utils/WellKnownUtils.ts [hunk 2]
```diff
@@ -70,6 +75,16 @@ export function tileServerFromWellKnown(
     );
 }
 
+export function getEmbeddedPagesWellKnown(): IEmbeddedPagesWellKnown | undefined {
+    return embeddedPagesFromWellKnown(MatrixClientPeg.get().getClientWellKnown());
```
Comment (t3chguy) [inline src/utils/WellKnownUtils.ts:79]:
```suggestion
    return embeddedPagesFromWellKnown(MatrixClientPeg.get()?.getClientWellKnown());
```

File: src/utils/WellKnownUtils.ts [hunk 3]
```diff
@@ -70,6 +75,16 @@ export function tileServerFromWellKnown(
...
+}
+
+export function embeddedPagesFromWellKnown(
+    clientWellKnown?: IClientWellKnown | undefined,
```
Comment (t3chguy) [inline src/utils/WellKnownUtils.ts:83]:
```suggestion
    clientWellKnown?: IClientWellKnown,
```

**Review by t3chguy — COMMENTED** [review]
Two tiny nits

File: src/utils/WellKnownUtils.ts (hunk 2 above)
Comment (johannes-krude) [reply src/utils/WellKnownUtils.ts:79]: I guess this is needed for future cases which might call this before logging in.

File: src/utils/WellKnownUtils.ts (hunk 3 above)
Comment (johannes-krude) [reply src/utils/WellKnownUtils.ts:83]: I am not familiar enough with typescript to understand why this is safe to do. FYI: `tileServerFromWellKnown` also accepts `undefined`.

**Comment by johannes-krude** [comment]
So I guess this can be merged. :tada: 

Thanks you for making this possible.

File: src/utils/WellKnownUtils.ts (hunk 2 above)
Comment (t3chguy) [reply src/utils/WellKnownUtils.ts]: The question mark makes an optional parameter

File: src/utils/WellKnownUtils.ts (hunk 3 above)
Comment (t3chguy) [reply src/utils/WellKnownUtils.ts]: The question mark makes it an optional parameter

**Pull request merged**